include/forcing/ForcingsEngine.hpp

```diff
@@ -0,0 +1,172 @@
+#pragma once
+
+#include <chrono>
+#include <memory>
+#include <vector>
+#include <boost/core/span.hpp>
+
+#include "Bmi_Py_Adapter.hpp"
+#include <utilities/mdarray.hpp>
+
+namespace data_access {
+
+struct ForcingsEngine
+{
+    using size_type                 = std::size_t;
+    using clock_type                = std::chrono::system_clock;
+    static constexpr auto bad_index = static_cast<size_type>(-1);
+
+    ForcingsEngine(const std::string& init, size_type time_start, size_type time_end);
+
+    ~ForcingsEngine();
+
+    /**
+     * @brief Get the output variables associated with this instance of the
+     *        Forcings Engine
+     * 
+     * @return boost::span<const std::string> 
+     */
+    boost::span<const std::string> outputs() const noexcept;
+
+    /**
+     * @brief Get the beginning time point for this instance of the
+     *        Forcings Engine
+     * 
+     * @return const clock_type::time_point& 
+     */
+    const clock_type::time_point& time_begin() const noexcept;
+
+    /**
+     * @brief Get the ending time point for this instance of the
+     *        Forcings Engine
+     * 
+     * @return const clock_type::time_point& 
+     */
+    const clock_type::time_point& time_end() const noexcept;
+
+    /**
+     * @brief Get the stride/time step this instance of the
+     *        Forcings Engine follows.
+     * 
+     * @return const clock_type::duration& 
+     */
+    const clock_type::duration& time_step() const noexcept;
+
+    /**
+     * @brief Get the index for a C-style epoch within this instance of the
+     *        Forcings Engine's temporal domain.
+     * 
+     * @param ctime C-style epoch
+     * @return size_type, where the index represents a point between time_begin() and time_end() by time_step().
+     */
+    size_type time_index(time_t ctime) const noexcept;
+
+    /**
+     * @brief Get the index in `CAT-ID` for a given divide in the instance cache.
+     * @note The `CAT-ID` output variable uses integer values instead of strings.
+     * 
+     * @param divide_id A hydrofabric divide ID, i.e. "cat-*"
+     * @return size_type 
+     */
+    size_type divide_index(const std::string& divide_id) const noexcept;
+
+    /**
+     * @brief Get the index of a variable in the instance cache.
+     * 
+     * @param variable 
+     * @return size_type 
+     */
+    size_type variable_index(const std::string& variable) const noexcept;
+
+    /**
+     * @brief Get a forcing value from the instance cache, or update the instance
+     *        and its cache.
```

Comment:
What does the "or update" refer to here?
- Is that a relic of an earlier iteration of the design?
- The return value isn't `double&`, so it can't be assigned to
- Does calling this result in the engine population the requested entry if it wasn't present in the cache?

Reply:
Broadly speaking, functions used to access data should probably not document their impact on any internal caches, as such. Caching is an implementation detail.

Reply:
Bad phrasing on my part, but it's the last bullet point there -- calling `at` with a later `raw_time` value will execute the model's `Update()` through to the new time and cache the values.